lightclient/src/platform/default.rs

```diff
@@ -0,0 +1,299 @@
+                    let write_queue_slices = buf.as_slices();
+                    if let Poll::Ready(result) = Pin::new(&mut stream.socket).poll_write_vectored(
+                        cx,
+                        &[
+                            IoSlice::new(write_queue_slices.0),
+                            IoSlice::new(write_queue_slices.1),
+                        ],
+                    ) {
+                        if !*must_close {
+                            // In the situation where the API user wants to close the writing
+                            // side, simply sending the buffered data isn't enough to justify
+                            // making the future ready.
+                            update_stream_future_ready = true;
+                        }
+
+                        match result {
+                            Err(_) => {
+                                // End the stream.
+                                stream.buffers = None;
+                                return Poll::Ready(());
+                            }
+                            Ok(bytes) => {
+                                *must_flush = true;
+                                for _ in 0..bytes {
+                                    buf.pop_front();
+                                }
+                            }
+                        }
+                    } else {
+                        break;
+                    }
+                }
+
+                if buf.is_empty() && *must_close {
+                    if let Poll::Ready(result) = Pin::new(&mut stream.socket).poll_close(cx) {
+                        update_stream_future_ready = true;
+                        match result {
+                            Err(_) => {
+                                // End the stream.
+                                stream.buffers = None;
+                                return Poll::Ready(());
+                            }
+                            Ok(()) => {
+                                *write_buffer = StreamWriteBuffer::Closed;
+                            }
+                        }
+                    }
+                } else if *must_flush {
+                    if let Poll::Ready(result) = Pin::new(&mut stream.socket).poll_flush(cx) {
+                        update_stream_future_ready = true;
+                        match result {
+                            Err(_) => {
+                                // End the stream.
+                                stream.buffers = None;
+                                return Poll::Ready(());
+                            }
+                            Ok(()) => {
+                                *must_flush = false;
+                            }
+                        }
+                    }
+                }
+            }
+
+            if update_stream_future_ready {
+                Poll::Ready(())
+            } else {
+                Poll::Pending
```

Comment:
Whenever `Poll::Pending` is returned, the `Context` (cx) _must_ at some point be used to wake the future up again, so that it continues to make progress. in at least one case, it looks like we return `Pending` here when nothing would be planning to ever call `wake()` on the context to wake it up again. Def something to ponder I think, so that things don't get stuck waiting forever :)

Reply:
Thankyou for adding the comments; as a reviewer it's scary to see the `Pending`'s without context wakers being used/stored/whatever, but I think I understand how it's working a bit better now!